vllm/model_executor/model_loader/loader.py

```diff
@@ -347,6 +347,124 @@ def load_model(self, *, model_config: ModelConfig,
+                        else load_config.model_loader_extra_config.copy())
+        self.pattern = extra_config.pop("pattern", self.DEFAULT_PATTERN)
+        if extra_config:
+            raise ValueError(f"Unexpected extra config keys for load format "
+                             f"{load_config.load_format}: "
+                             f"{load_config.model_loader_extra_config.keys()}")
+
+    @staticmethod
+    def _filter_subtensors(tensors: Dict[str, torch.Tensor]):
+        """
+        Filter out all tensors that share the same memory or a subset of the
+        memory of another tensor.
+        """
+        from safetensors.torch import storage_ptr, storage_size
+        result = {}
+        for key1, tensor1 in tensors.items():
+            a1 = storage_ptr(tensor1)  # tensor1 start
+            b1 = a1 + storage_size(tensor1)  # tensor1 end
+            for key2, tensor2 in tensors.items():
+                a2 = storage_ptr(tensor2)  # tensor2 start
+                b2 = a2 + storage_size(tensor2)  #tensor2 end
+                if (a1, b1) == (a2, b2):
+                    # Same memory, take only the first key (lexicographically).
+                    if key2 < key1:
+                        break
+                elif a1 <= a2 and b2 <= b1:
+                    # tensor1 is a subtensor of tensor2.
+                    break
+            else:
+                result[key1] = tensor1
+        return result
+
+    def load_model(self, *, model_config: ModelConfig,
+                   device_config: DeviceConfig,
+                   lora_config: Optional[LoRAConfig],
+                   vision_language_config: Optional[VisionLanguageConfig],
+                   parallel_config: ParallelConfig,
+                   scheduler_config: SchedulerConfig) -> nn.Module:
+        from safetensors.torch import load_file
+
+        from vllm.distributed import get_tensor_model_parallel_rank
+        with set_default_torch_dtype(model_config.dtype):
+            with torch.device(device_config.device):
+                model = _initialize_model(model_config, self.load_config,
+                                          lora_config, vision_language_config)
+            rank = get_tensor_model_parallel_rank()
+            pattern = os.path.join(
+                model_config.model,
+                self.pattern.format(rank=rank, part="*"),
+            )
+            filepaths = glob.glob(pattern)
+            if not filepaths:
+                # TODO: support un-sharded checkpoints too
+                raise ValueError(
+                    f"Could not find checkpoint files '{pattern}', only "
+                    f"pre-sharded checkpoints are currently supported!"
+                )
+            state_dict = self._filter_subtensors(model.state_dict())
+            for path in filepaths:
+                for key, tensor in load_file(path).items():
+                    state_dict[key].copy_(tensor)
+                    state_dict.pop(key)
```

Comment:
this will load the entire file eagerly, better approach should be:
```suggestion
            for path in filepaths:
                with safe_open(filename, framework="pt", device=device) as f:
                    for key in state_dict:
                        state_dict[key].copy_(f.get_tensor(key))
```